### DISCUSSION TITLE
Zombie clients

### DISCUSSION THREAD

**Discussion by mpwyn** [ltsp/ltsp#658] [General] [answered]
We are still in the process of migrating from ltsp5 to ltsp22. We've made a lot of progress and are very impressed by all the improvements in ltsp22, so congratulations to Alkis for a job very well done!

We are using the chroot approach and thus use "ltsp image" to create an .img file. We currently have clients that successfully boot using the .img file.

While the clients are running we continue to make modifications in the chroot, and regenerate the .img file every now and then. Every time we regenerate the .img file the old version gets renamed to .img.old. After the second regeneration the original .img file - the one that the currently running clients booted from - is deleted.

This is no problem in ltsp5 using nbd because the daemon doesn't release the .img file's inodes until the client disconnects.  This doesn't seem to be the case in ltsp22 in which the client uses nfs to access the .img file. What happens is that the client enters some sort of zombie state, half functional and half non-functional.

Is our understanding of what is going on correct? Is the .img file really completely lost when the .img is regenerated twice after the client boots? That would be very restrictive for our installation because we have clients that in extreme cases run up to two months without being restarted.

Today we succeeded in setting the site up to use nbd but we would prefer to use nfs since it's the standard for ltsp22.

Matthew

**Comment by alkisg** [comment, marked as answer]
Hi Matthew, thank you for your good words,
you described the issue perfectly, I also wrote about it in https://github.com/ltsp/ltsp/issues/132#issuecomment-1000381677.
So currently LTSP is good for only one run of `ltsp image`; for a second run, the clients will need to be rebooted in the meantime.
I do mention a workaround there, to rename the ltsp images using their dates and to avoid deleting the ones that are up to XXX days old.
So that would be your official site policy, that "people should reboot e.g. once per month; otherwise your terminal will hang".

The other workaround would be to write a POST_SERVICE_xxx line that mounts the image using normal NFS locks in a dummy place, just so that it's in use and can't be deleted from the server.

The proper solution of locking the actual mounts of course can't be offered by LTSP but needs to be implemented in initramfs-tools, affected users would need to file an issue there.
